services/rollout-service/pkg/service/broadcast.go

```diff
@@ -33,38 +34,101 @@ type key struct {
 	Environment string
 }
 
+type appState struct {
+	argocdVersion    uint64
+	kuberpultVersion uint64
+	rolloutStatus    api.RolloutStatus
+}
+
+func (a *appState) applyArgoEvent(ev *ArgoEvent) *BroadcastEvent {
+	status := rolloutStatus(ev)
+	if a.rolloutStatus != status || a.argocdVersion != ev.Version {
+		a.rolloutStatus = status
+		a.argocdVersion = ev.Version
+		return a.getEvent(ev.Application, ev.Environment)
+	}
+	return nil
+}
+
+func (a *appState) applyKuberpultEvent(ev *versions.KuberpultEvent) *BroadcastEvent {
+	if a.kuberpultVersion != ev.Version {
+		a.kuberpultVersion = ev.Version
+		return a.getEvent(ev.Application, ev.Environment)
+	}
+	return nil
+}
+
+func (a *appState) getEvent(application, environment string) *BroadcastEvent {
+	rs := a.rolloutStatus
+	if a.kuberpultVersion != 0 && a.kuberpultVersion != a.argocdVersion {
+		rs = api.RolloutStatus_RolloutStatusProgressing
+	}
+	return &BroadcastEvent{
+		Environment:      environment,
+		Application:      application,
+		ArgocdVersion:    a.argocdVersion,
+		RolloutStatus:    rs,
+		KuberpultVersion: a.kuberpultVersion,
+	}
+}
+
 type Broadcast struct {
-	state    map[key]Event
+	state    map[key]*appState
```

Comment:
We still do need to lock things because we also have to consider the subscribers to that state AND we also need to merge the state of argocd and kuberpult. So even if we could have two seperate stores that would not require locking we would still need to have a common state that needs locking.

Reply:
Follow up: https://github.com/freiheit-com/kuberpult/issues/915